storm-client/src/jvm/org/apache/storm/topology/PersistentWindowedBoltExecutor.java

```diff
@@ -0,0 +1,596 @@
+                    }
+                );
+        }
+
+        private void deletePartition(long pid) {
+            LOG.debug("Delete partition: {}", pid);
+            try {
+                partitionIdsLock.lock();
+                partitionIds.remove(pid);
+                partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
+            } finally {
+                partitionIdsLock.unlock();
+            }
+        }
+
+        private long getNextPartitionId() {
+            try {
+                partitionIdsLock.lock();
+                partitionIds.add(++latestPartitionId);
+                partitionIdsState.put(PARTITION_IDS_KEY, partitionIds);
+            } finally {
+                partitionIdsLock.unlock();
+            }
+            return latestPartitionId;
+        }
+
+        private WindowPartition<T> getPinnedPartition(long id) {
+            return cache.getPinned(id);
+        }
+
+        private void flush() {
+            LOG.debug("Flushing modified partitions");
+            cache.asMap().forEach((pid, p) -> {
+                Long pidToInvalidate = null;
+                try {
+                    windowPartitionsLock.lock();
+                    if (p.isEmpty() && pid != latestPartitionId) {
+                        LOG.debug("Invalidating empty partition {}", pid);
+                        deletePartition(pid);
+                        windowPartitionsState.delete(pid);
+                        pidToInvalidate = pid;
+                    } else if (p.isModified()) {
+                        LOG.debug("Updating modified partition {}", pid);
+                        p.clearModified();
+                        windowPartitionsState.put(pid, p);
+                    }
+                } finally {
+                    windowPartitionsLock.unlock();
+                }
+                // invalidate after releasing the lock
+                if (pidToInvalidate != null) {
+                    cache.invalidate(pidToInvalidate);
+                }
+            });
+            windowSystemStateSupplier.get().forEach(windowSystemState::put);
+        }
+    }
+
+    // the window partition that holds the events
+    public static class WindowPartition<T> implements Iterable<Event<T>> {
+        private final ConcurrentLinkedQueue<Event<T>> events = new ConcurrentLinkedQueue<>();
+        private final AtomicInteger size = new AtomicInteger();
+        private final long id;
+        private transient volatile boolean modified;
+
+        public WindowPartition(long id) {
+            this.id = id;
+        }
+
+        void add(Event<T> event) {
+            events.add(event);
+            size.incrementAndGet();
+            setModified();
+        }
+
+        boolean isModified() {
+            return modified;
+        }
+
+        void setModified() {
+            if (!modified) {
+                modified = true;
+            }
+        }
+
+        void clearModified() {
+            if (modified) {
```

Comment:
In `clearModified` its probably ok to skip the check since its invoked only during flush. 

`setModified` is invoked in `add` so we can use it to avoid frequent writes to the volatile.